### PR TITLE
[ISTHMUS]support order by, limit offset clause in substrait to sql conversion

### PR DESCRIPTION
**Pull request by weijietong**
fix issue #40

### CODE REVIEW

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java [hunk 1]
```diff
@@ -100,6 +107,57 @@ public RelNode visit(Project project) throws RuntimeException {
     return relBuilder.push(child).project(rexList).build();
   }
 
+  @Override
+  public RelNode visit(Sort sort) throws RuntimeException {
+    RelNode child = sort.getInput().accept(this);
+    List<RelFieldCollation> relFieldCollations =
+        sort.getSortFields().stream().map(sortField -> toRelFieldCollation(sortField)).toList();
+    if (relFieldCollations.isEmpty()) {
```
Comment (JamesRTaylor) [inline isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:115]: Is this if statement necessary?

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java [hunk 2]
```diff
@@ -100,6 +107,57 @@ public RelNode visit(Project project) throws RuntimeException {
...
+        nullDirection = RelFieldCollation.NullDirection.FIRST;
+        fieldDirection = RelFieldCollation.Direction.DESCENDING;
+      }
+      case DESC_NULLS_LAST -> {
+        nullDirection = RelFieldCollation.NullDirection.LAST;
+        fieldDirection = RelFieldCollation.Direction.DESCENDING;
+      }
+      case CLUSTERED -> fieldDirection = RelFieldCollation.Direction.CLUSTERED;
+
+      default -> throw new RuntimeException("Unexpected Expression.SortDirection enum!");
```
Comment (JamesRTaylor) [inline isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:156]: Add sortDirection to exception message

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java [hunk 3]
```diff
@@ -100,6 +107,57 @@ public RelNode visit(Project project) throws RuntimeException {
...
+    }
+    return relBuilder.push(child).sort(RelCollations.of(relFieldCollations)).build();
+  }
+
+  @Override
+  public RelNode visit(Fetch fetch) throws RuntimeException {
+    RelNode child = fetch.getInput().accept(this);
+    var optCount = fetch.getCount();
+    long count = -1L;
+    if (optCount.isPresent()) {
```
Comment (jacques-n) [inline isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:126]: You can avoid all of this using `optCount.orElse(-1L)`

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java [hunk 4]
```diff
@@ -100,6 +107,57 @@ public RelNode visit(Project project) throws RuntimeException {
...
+      count = optCount.getAsLong();
+    }
+    var offset = fetch.getOffset();
+    return relBuilder.push(child).limit((int) offset, (int) count).build();
```
Comment (jacques-n) [inline isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:130]: Let's put a check above this that throws if these truncations are going to cause a lost of value (as opposed to silently truncating).

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java (hunk 1 above)
Comment (weijietong) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:115]: yes, it is to solve the `select xxx from xxx limit 100` case which does not have the `order by` clause

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java (hunk 2 above)
Comment (weijietong) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java]: done

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java (hunk 3 above)
Comment (weijietong) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java]: that's elegant! done.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java (hunk 4 above)
Comment (weijietong) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:250]: done

**Review by JamesRTaylor — APPROVED** [review]

**Comment by jacques-n** [comment]
This needs to be rebased before merging.

**Comment by JamesRTaylor** [comment]
LGTM. Thanks for the fix, @weijietong.

**Pull request merged**